File: internal/render/render.go
package render

import (
	"bytes"
	"embed"
	"fmt"
	"html/template"
	"io/fs"

	"mdslides/internal/layout"
	"mdslides/internal/markdown"
)

//go:embed web/css web/js
var staticFiles embed.FS

// StaticAssets holds the embedded CSS/JS the viewer needs, rooted so that
// StaticAssets's "css/tokens.css" is web/css/tokens.css. The server mounts
// this under /static/ without needing to know anything about what's
// inside — HTTP routing doesn't require HTML/CSS/JS knowledge. Embedding
// the two directories (rather than listing every file) means a new CSS
// or JS file under web/css or web/js is picked up automatically, no
// go:embed edit needed.
var StaticAssets = mustSub(staticFiles, "web")

func mustSub(f embed.FS, dir string) fs.FS {
	sub, err := fs.Sub(f, dir)
	if err != nil {
		// Only reachable if the embed path above is wrong, which is a
		// compile-time-guaranteed-correct constant, not runtime input —
		// panicking here means "this is a bug in this package," the same
		// idiom as template.Must below.
		panic(err)
	}
	return sub
}

//go:embed web/page.html.tmpl
var templateFS embed.FS

var pageTemplate = template.Must(template.ParseFS(templateFS, "web/page.html.tmpl"))

// screen is the per-<section> data the template ranges over. The type
// itself doesn't need to be exported (only this package ever builds one),
// but its fields do — html/template reads them by reflection, which
// requires exported fields regardless of the type's own visibility.
type screen struct {
	Heading     template.HTML
	ContentHTML template.HTML
	Images      []markdown.Image
	Layout      string
	ImagesFirst bool
}

type document struct {
	Title   string
	Screens []screen
}

// HTML renders a full HTML document for the deck: every Slide's Pages
// become one or more <section> "screens". A Page with more images than
// layout.MaxPerPage is split into extra images-only continuation screens
// rather than overflowing one grid — see screensForPage.
func HTML(deck markdown.Deck) (string, error) {
	doc := document{Title: deck.Title}
	for _, slide := range deck.Slides {
		for _, page := range slide.Pages {
			doc.Screens = append(doc.Screens, screensForPage(slide.HeadingHTML, page)...)
		}
	}

	var buf bytes.Buffer
	if err := pageTemplate.Execute(&buf, doc); err != nil {
		return "", fmt.Errorf("execute page template: %w", err)
	}
	return buf.String(), nil
}

// screensForPage turns one markdown.Page into one or more rendered
// screens. A page within layout.MaxPerPage images is exactly one screen;
// beyond that, the extra images become their own images-only continuation
// screens, heading repeated with " (cont.)" and no text — repeating the
// same paragraph on every continuation screen was more confusing than
// just showing more images under the same heading.
func screensForPage(heading template.HTML, page markdown.Page) []screen {
	chunks := chunkImages(page.Images, layout.MaxPerPage)
	if len(chunks) == 0 {
		chunks = [][]markdown.Image{nil}
	}

	screens := make([]screen, 0, len(chunks))
	for i, images := range chunks {
		s := screen{
			Heading:     heading,
			Layout:      layout.Classify(len(images)).String(),
			Images:      images,
			ImagesFirst: page.ImagesFirst,
		}
		if i == 0 {
			s.ContentHTML = page.ContentHTML
		} else {
			s.Heading = heading + template.HTML(" (cont.)")
		}
		screens = append(screens, s)
	}
	return screens
}

// chunkImages splits images into groups of at most size, preserving
// order; an empty input or a non-positive size yields no chunks. The
// [:n:n] full slice expression caps each chunk's capacity at its length,
// so nothing downstream can accidentally grow one chunk into the next
// image's memory via append.
func chunkImages(images []markdown.Image, size int) [][]markdown.Image {
	if len(images) == 0 || size <= 0 {
		return nil
	}
	var chunks [][]markdown.Image
	for len(images) > 0 {
		n := size
		if n > len(images) {
			n = len(images)
		}
		chunks = append(chunks, images[:n:n])
		images = images[n:]
	}
	return chunks
}

File: extension/src/deck/render.ts
// Port of internal/render/render.go plus the <section> half of
// web/page.html.tmpl. The markup is byte-for-byte the shape the shared CSS
// expects: .screen.layout-<kind>[.images-first] > .screen-heading +
// .screen-body > .screen-content + .screen-images.

import { classify, MAX_PER_PAGE, type LayoutKind } from "./layout";
import type { Deck, Image, Page } from "./model";

/** One screen the viewer shows: the unit nav.js pages through. */
export interface Screen {
  /** Heading HTML, inline formatting preserved. */
  heading: string;
  contentHtml: string;
  images: Image[];
  layout: LayoutKind;
  imagesFirst: boolean;
  /** 0-based source line, for mapping the editor cursor onto a screen. */
  line: number;
}

export interface RenderedDeck {
  title: string;
  screens: Screen[];
  html: string;
}

/**
 * Flattens a Deck into screens and renders them. Every Slide's Pages become
 * one or more screens; a Page with more images than MAX_PER_PAGE is split into
 * extra images-only continuation screens rather than overflowing one grid.
 */
export function renderDeck(deck: Deck): RenderedDeck {
  const screens: Screen[] = [];
  for (const slide of deck.slides) {
    for (const page of slide.pages) {
      screens.push(...screensForPage(slide.headingHtml, page));
    }
  }
  return {
    title: deck.title,
    screens,
    html: screens.map(renderScreen).join("\n"),
  };
}

/**
 * Turns one Page into one or more screens. A page within MAX_PER_PAGE images
 * is exactly one screen; beyond that the extra images become their own
 * images-only continuation screens, heading repeated with " (cont.)" and no
 * text — repeating the same paragraph on every continuation screen was more
 * confusing than just showing more images under the same heading.
 */
function screensForPage(heading: string, page: Page): Screen[] {
  let chunks = chunkImages(page.images, MAX_PER_PAGE);
  if (chunks.length === 0) {
    chunks = [[]];
  }

  return chunks.map((images, i) => ({
    heading: i === 0 ? heading : `${heading} (cont.)`,
    contentHtml: i === 0 ? page.contentHtml : "",
    images,
    layout: classify(images.length),
    imagesFirst: page.imagesFirst,
    line: page.line,
  }));
}

/** Splits images into groups of at most size, preserving order. */
function chunkImages(images: Image[], size: number): Image[][] {
  if (images.length === 0 || size <= 0) {
    return [];
  }
  const chunks: Image[][] = [];
  for (let i = 0; i < images.length; i += size) {
    chunks.push(images.slice(i, i + size));
  }
  return chunks;
}

function renderScreen(screen: Screen, index: number): string {
  const classes = `screen layout-${screen.layout}${screen.imagesFirst ? " images-first" : ""}`;
  const parts: string[] = [];
  parts.push(`<section class="${classes}" data-index="${index}" data-line="${screen.line}">`);
  if (screen.heading) {
    parts.push(`<h1 class="screen-heading">${screen.heading}</h1>`);
  }
  parts.push('<div class="screen-body">');
  if (screen.contentHtml) {
    parts.push(`<div class="screen-content">${screen.contentHtml}</div>`);
  }
  if (screen.images.length > 0) {
    const imgs = screen.images
      .map((img) => `<img src="${attr(safeUrl(img.src))}" alt="${attr(img.alt)}">`)
      .join("");
    parts.push(`<div class="screen-images">${imgs}</div>`);
  }
  parts.push("</div>");
  parts.push("</section>");
  return parts.join("\n");
}

function attr(value: string): string {
  return value
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

/**
 * Blanks out URL schemes that execute code. Go's html/template does this for
 * free when it sees a value land in an href/src attribute; building the markup
 * by hand here means doing it explicitly, and a deck is rendered from whatever
 * Markdown happens to be open in the editor.
 */
function safeUrl(src: string): string {
  if (/^\s*(javascript|vbscript):/i.test(src)) {
    return "";
  }
  return src;
}
